In [137]:
import pandas as pd
dataset = pd.read_csv("general_data.csv")

In [138]:
dataset.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [139]:
dataset.describe(include="all")

Age Attrition BusinessTravel              Department  \
count   4410.000000      4410           4410                    4410   
unique          NaN         2              3                       3   
top             NaN        No  Travel_Rarely  Research & Development   
freq            NaN      3699           3129                    2883   
mean      36.923810       NaN            NaN                     NaN   
std        9.133301       NaN            NaN                     NaN   
min       18.000000       NaN            NaN                     NaN   
25%       30.000000       NaN            NaN                     NaN   
50%       36.000000       NaN            NaN                     NaN   
75%       43.000000       NaN            NaN                     NaN   
max       60.000000       NaN            NaN                     NaN   

        DistanceFromHome    Education EducationField  EmployeeCount  \
count        4410.000000  4410.000000           4410         4410.0   
unique               NaN          NaN              6            NaN   
top                  NaN          NaN  Life Sciences            NaN   
freq                 NaN          NaN           1818            NaN   
mean            9.192517     2.912925            NaN            1.0   
std             8.105026     1.023933            NaN            0.0   
min             1.000000     1.000000            NaN            1.0   
25%             2.000000     2.000000            NaN            1.0   
50%             7.000000     3.000000            NaN            1.0   
75%            14.000000     4.000000            NaN            1.0   
max            29.000000     5.000000            NaN            1.0   

         EmployeeID Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  \
count   4410.000000   4410  ...         4391.000000   4410       4410.000000   
unique          NaN      2  ...                 NaN      1               NaN   
top             NaN   Male  ...                 NaN      Y               NaN   
freq            NaN   2646  ...                 NaN   4410               NaN   
mean    2205.500000    NaN  ...            2.694830    NaN         15.209524   
std     1273.201673    NaN  ...            2.498887    NaN          3.659108   
min        1.000000    NaN  ...            0.000000    NaN         11.000000   
25%     1103.250000    NaN  ...            1.000000    NaN         12.000000   
50%     2205.500000    NaN  ...            2.000000    NaN         14.000000   
75%     3307.750000    NaN  ...            4.000000    NaN         18.000000   
max     4410.000000    NaN  ...            9.000000    NaN         25.000000   

        StandardHours  StockOptionLevel TotalWorkingYears  \
count          4410.0       4410.000000       4401.000000   
unique            NaN               NaN               NaN   
top               NaN               NaN               NaN   
freq              NaN               NaN               NaN   
mean              8.0          0.793878         11.279936   
std               0.0          0.851883          7.782222   
min               8.0          0.000000          0.000000   
25%               8.0          0.000000          6.000000   
50%               8.0          1.000000         10.000000   
75%               8.0          1.000000         15.000000   
max               8.0          3.000000         40.000000   

        TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
count             4410.000000     4410.000000              4410.000000   
unique                    NaN             NaN                      NaN   
top                       NaN             NaN                      NaN   
freq                      NaN             NaN                      NaN   
mean                 2.799320        7.008163                 2.187755   
std                  1.288978        6.125135                 3.221699   
min                  0.000000        0.000000                 0.000000   
25%         

In [140]:
dataset.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

In [141]:
dataset.isnull

<bound method DataFrame.isnull of       Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
1      31       Yes  Travel_Frequently  Research & Development   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   
4409   40        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
1              

In [142]:
dataset.duplicated() 


0       False
1       False
2       False
3       False
4       False
        ...  
4405    False
4406    False
4407    False
4408    False
4409    False
Length: 4410, dtype: bool

In [143]:
dataset.drop_duplicates()


Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
1      31       Yes  Travel_Frequently  Research & Development   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   
4409   40        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
1                   10          1  Life Sciences              1           2   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
...                ...        ...            ...            ...         ...   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   
4409                28          3        Medical              1        4410   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
1     Female  ...                 0.0      Y                23              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
...      ...  ...                 ...    ...               ...            ...   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   
4409    Male  ...                 0.0      Y                12              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
1                    1               6.0                      3   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
...                ...               ...                    ...   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   
4409                 0               NaN                      6   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
1                  5                        1                     4  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
...              ...                      ...                   ..

In [144]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      4410 non-null   int64  
 1   Attrition                4410 non-null   object 
 2   BusinessTravel           4410 non-null   object 
 3   Department               4410 non-null   object 
 4   DistanceFromHome         4410 non-null   int64  
 5   Education                4410 non-null   int64  
 6   EducationField           4410 non-null   object 
 7   EmployeeCount            4410 non-null   int64  
 8   EmployeeID               4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

## Step 5 – Statistical Tests (Mann-Whitney)


### Attrition Vs Distance from Home 

In [145]:
from sklearn import preprocessing
Label_encoder = preprocessing.LabelBinarizer()
dataset.Attrition = Label_encoder.fit_transform(dataset.Attrition)
dataset.Attrition


0       0
1       1
2       0
3       0
4       0
       ..
4405    0
4406    0
4407    0
4408    0
4409    0
Name: Attrition, Length: 4410, dtype: int32

In [146]:
import pandas as pd
datasetYes = pd.read_excel("Attrition_Yes_No.xls", sheet_name="Yes")
datasetYes.head()


Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   31       Yes  Travel_Frequently  Research & Development                10   
1   28       Yes      Travel_Rarely  Research & Development                11   
2   47       Yes         Non-Travel  Research & Development                 1   
3   44       Yes  Travel_Frequently  Research & Development                 1   
4   26       Yes      Travel_Rarely  Research & Development                 4   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          1  Life Sciences              1           2  Female  ...   
1          2        Medical              1           7    Male  ...   
2          1        Medical              1          14    Male  ...   
3          2        Medical              1          29    Male  ...   
4          3        Medical              1          31    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 0.0      Y                23              8   
1                 2.0      Y                20              8   
2                 1.0      Y                11              8   
3                 3.0      Y                14              8   
4                 2.0      Y                11              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 1               6.0                      3               5   
1                 1               5.0                      2               0   
2                 2              10.0                      4              10   
3                 1              19.0                      2               1   
4                 0               5.0                      5               3   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        1                     4  
1                        0                     0  
2                        9                     9  
3                        0                     0  
4                        0                     2  

[5 rows x 24 columns]

In [147]:
datasetNo = pd.read_excel("Attrition_Yes_No.xls", sheet_name= "No")
datasetNo.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   32        No  Travel_Frequently  Research & Development                17   
2   38        No         Non-Travel  Research & Development                 2   
3   32        No      Travel_Rarely  Research & Development                10   
4   46        No      Travel_Rarely  Research & Development                 8   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          4          Other              1           3    Male  ...   
2          5  Life Sciences              1           4    Male  ...   
3          1        Medical              1           5    Male  ...   
4          3  Life Sciences              1           6  Female  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 1.0      Y                15              8   
2                 3.0      Y                11              8   
3                 4.0      Y                12              8   
4                 3.0      Y                13              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 3               5.0                      2               5   
2                 3              13.0                      5               8   
3                 2               9.0                      2               6   
4                 0              28.0                      5               7   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        0                     3  
2                        7                     5  
3                        0                     4  
4                        7                     7  

[5 rows x 24 columns]

In [148]:
from scipy.stats import mannwhitneyu  


In [149]:
a1 = datasetYes.DistanceFromHome
a2 = datasetNo.DistanceFromHome
stat,p = mannwhitneyu(a1,a2)
print(stat,p)


1312110.0 0.4629185205822659


#### H0: There is no significant differences in the Distance From Home between attrition (Y) and attirition (N) 
#### Ha: There is significant differences in the Distance From Home between attrition (Y) and attirition (N) 
#### As the P value of 0.46 is > 0.05, the Ha is rejected and H0 is accepted.  

### Attrition Vs Income 

In [150]:
a1 = datasetYes.MonthlyIncome
a2 = datasetNo.MonthlyIncome
stat,p = mannwhitneyu(a1,a2)
print(stat,p)

1264900.5 0.053577283839938566


#### As the P value is 0.0536, which is > than 0.05, the HA is rejected and h0 is accepted.  
#### H0: There is no significant differences in the income between attrition (Y) and attirition (N) 
#### Ha: There is significant differences in the income between attrition (Y) and attirition (N) 

### Attrition Vs Total Working Years

In [151]:
a1 = datasetYes.TotalWorkingYears
a2 = datasetNo.TotalWorkingYears
stat,p = mannwhitneyu(a1,a2)
print(stat,p)

909954.0 2.8897554531400447e-39


#### As the P value is 2.8897554531400447e-39, which is > than 0.05, the HA is rejected and H0 is accepted.  
#### H0: There is no significant differences in the Total Working Years between attrition (Y) and attirition (N) 
#### HA: There is significant differences in the Total Working Years between attrition (Y) and attirition (N) 

### Attrition Vs Years at Company 

In [152]:
a1 = datasetYes.YearsAtCompany
a2 = datasetNo.YearsAtCompany
stat,p = mannwhitneyu(a1,a2)
print(stat,p)

923238.0 6.047598261692858e-37


#### As the P value is 6.047598261692858e-37, which is < than 0.05, the H0 is rejected and HA is accepted.  
#### H0: There is no significant differences in the Years At Company between attrition (Y) and attirition (N) 
####  HA: There is significant differences in the Years At Company between attrition (Y) and attirition (N) 

### Attrition Vs YearsWithCurrentManager

In [153]:
a1 = datasetYes.YearsWithCurrManager
a2 = datasetNo.YearsWithCurrManager
stat,p = mannwhitneyu(a1,a2)
print(stat,p)

957253.5 1.2365483142169853e-31


#### As the P value is 1.2365483142169853e-31, which is < than 0.05, the H0 is rejected and HA is accepted.  
#### H0: There is no significant differences in the Years With Current Manager between attrition (Y) and attirition (N) 
####  Ha: There is significant differences in the Years With Current Manager between attrition (Y) and attirition (N) 

## Statistical Tests (Separate T Test) 

In [154]:
from scipy.stats import ttest_ind

### Attrition Vs Distance From Home

In [155]:
z1 = datasetYes.DistanceFromHome
z2 = datasetNo.DistanceFromHome
stat,p = ttest_ind(z2,z1)
print(stat,p)

0.6460416038042738 0.518286042805572


#### As the P value is 0.518286042805572, which is > than 0.05, the HA is rejected and H0 is accepted.  
#### H0: There is no significant differences in the Distance From Home between attrition (Y) and attirition (N) 
####  HA: There is significant differences in the Distance From Home between attrition (Y) and attirition (N) 

### Attrition Vs Income 

In [156]:
z1 = datasetYes.MonthlyIncome
z2 = datasetNo.MonthlyIncome
stat,p = ttest_ind(z2,z1)
print(stat,p)

2.0708863763619316 0.03842748490605113


#### As the P value is 0.03842748490605113, which is < than 0.05, the HA is rejected and H0 is accepted.  
#### H0: There is no significant differences in the Monthly Income between attrition (Y) and attirition (N) 
#### HA: There is significant differences in the Monthly Income between attrition (Y) and attirition (N) 

### Attrition Vs Yeats At Company

In [157]:
z1 = datasetYes.MonthlyIncome
z2 = datasetNo.MonthlyIncome
stat,p = ttest_ind(z2,z1)
print(stat,p)

2.0708863763619316 0.03842748490605113


#### As the P value is 0.03842748490605113, which is > than 0.05, the HA is rejected and H0 is accepted.  
#### H0: There is no significant differences in the Years At Company between attrition (Y) and attirition (N) 
#### HA: There is significant differences in the Years At Company between attrition (Y) and attirition (N) 

### Attrition Vs Years With Current Manager 

In [158]:
z1 = datasetYes.YearsWithCurrManager
z2 = datasetNo.YearsWithCurrManager
stat,p = ttest_ind(z2,z1)
print(stat,p)

10.499379408703438 1.7339322652918153e-25


#### As the P value is 1.7339322652918153e-25, which is < than 0.05, the H0 is rejected and HA is accepted.  
#### H0: There is no significant differences in the Years With Current Manager between attrition (Y) and attirition (N) 
#### HA: There is significant differences in the Years With Current Manager between attrition (Y) and attirition (N) 

## Unsupervised Learning - Correlation Analysis 

In [159]:
from scipy.stats import pearsonr 

In [160]:
dataset['TotalWorkingYears']=dataset['TotalWorkingYears'].fillna(11.28) 

In [161]:
stats, p=pearsonr(dataset.Attrition, dataset.DistanceFromHome) 
print(stats, p) 

-0.00973014101017966 0.5182860428050771


In [162]:
stats, p=pearsonr(dataset.Attrition, dataset.MonthlyIncome) 
print(stats, p) 

-0.031176281698115007 0.03842748490600132


In [163]:
stats, p=pearsonr(dataset.Attrition, dataset.TotalWorkingYears) 
print(stats, p) 

-0.17011136355965176 5.473159751791005e-30


In [164]:
stats, p=pearsonr(dataset.Attrition, dataset.YearsAtCompany) 
print(stats, p) 

-0.1343922139899772 3.1638831224877484e-19


In [165]:
stats, p=pearsonr(dataset.Attrition, dataset.YearsWithCurrManager) 
print(stats, p) 

-0.15619931590162842 1.7339322652900218e-25


### The inference of the above analysis are as follows: 

### Attrition & DistanceFromHome:  
#### As r = -0.009, there’s low negative correlation between Attrition and DistanceFromHome 
#### As the P value of 0.518 is > 0.05, we are accepting H0 and hence there’s no significant correlation 
#### between Attrition & DistanceFromHome 

### Attrition & MonthlyIncome: 
#### As r = -0.031, there’s low negative correlation between Attrition and MonthlyIncome 
#### As the P value of 0.038 is < 0.05, we are accepting Ha and hence there’s significant correlation 
#### between Attrition & MonthlyIncome 

### Attrition & TotalWorkingYears: 
#### As r = -0.17, there’s low negative correlation between Attrition and TotalWorkingYears 
#### As the P value is < 0.05, we are accepting Ha and hence there’s significant correlation between 
#### Attrition & TotalWorkingYears 

### Attrition & YearsAtCompany: 
#### As r = -0.1343, there’s low negative correlation between Attrition and YearsAtCompany 
#### As the P value is < 0.05, we are accepting Ha and hence there’s significant correlation between 
#### Attrition & YearsAtCompany 

### Attrition & YearsWithCurrManager: 
#### As r = -0.1561, there’s low negative correlation between Attrition and YearsWithCurrManager 
#### As the P value is < 0.05, we are accepting Ha and hence there’s significant correlation between 
#### Attrition & YearsWithCurrManager    